XenCenterLib/Archive/ArchiveWriter.cs

```diff
@@ -73,59 +73,40 @@ public void CreateArchive(string pathToArchive, Action cancellingDelegate = null
-                throw new FileNotFoundException("The path " + pathToArchive + " does not exist");
+            pathToCurrentDirectory = StringUtility.ToLongWindowsPath(pathToCurrentDirectory, true);
 
-            AddFiles(longPathToArchive, pathToCurrentDirectory, cancellingDelegate);
+            //add the current directory except when it's the root directory
+            if (pathToArchive != pathToCurrentDirectory)
+                AddDirectory(CleanRelativePathName(pathToArchive, pathToCurrentDirectory), Directory.GetCreationTime(pathToCurrentDirectory));
 
+            var files = Directory.GetFiles(pathToCurrentDirectory);
             var directories = Directory.GetDirectories(pathToCurrentDirectory);
+            float increment = totalProgressIncrease / (files.Length + directories.Length);
+            float currentPercentage = progressOffset;
 
-            if (directories.Length == 0)
+            foreach (var file in files)
             {
-                progressDelegate?.Invoke(totalProgressIncrease + progressOffset);
-                return;
-            }
-
-            var progressIncrement = totalProgressIncrease / directories.Length;
-            
-            for (var j = 0; j < directories.Length; j++)
-            {
-                var subdirectoryPath = directories[j];
-                
-                AddDirectory(CleanRelativePathName(pathToArchive, subdirectoryPath), Directory.GetCreationTime(subdirectoryPath));
-
-                // Add the remainder to the offset of the last subdirectory
-                var remainder = j == directories.Length - 1 ? totalProgressIncrease % directories.Length : 0;
-                var newProgressOffset = j * progressIncrement + progressOffset + remainder;
+                cancellingDelegate?.Invoke();
 
-                PopulateArchive(pathToArchive, subdirectoryPath, cancellingDelegate, progressDelegate,progressIncrement, newProgressOffset);
-            }
-        }
+                var filePath = StringUtility.ToLongWindowsPath(file, false);
 
-        private void AddFiles(string pathToArchive, string currentDirectory, Action cancellingDelegate)
-        {
-            var files = Directory.GetFiles(StringUtility.ToLongWindowsPath(currentDirectory));
-            foreach (var filePath in files)
-            {
-                cancellingDelegate?.Invoke();
+                using (var fs = File.OpenRead(filePath))
+                    Add(fs, CleanRelativePathName(pathToArchive, filePath), File.GetCreationTime(filePath), cancellingDelegate);
 
-                using (var fs = File.OpenRead(StringUtility.ToLongWindowsPath(filePath)))
-                {
-                    Add(fs, CleanRelativePathName(pathToArchive, filePath), File.GetCreationTime(StringUtility.ToLongWindowsPath(filePath)), cancellingDelegate);
-                }
+                currentPercentage += increment;
+                progressDelegate?.Invoke((int)currentPercentage);

```

Comment:
I'm afraid that like this we're losing the same level of accuracy when reporting progress. Even ignoring remainders (which to be fair are not really needed)

Firstly files closer to the root directory have a disproportionately higher influence on the total progress.

Secondly, in `var directory in directories` we are not giving subdirectories any offset.

so with a tree like this:

```
root
├───dir1
├───dir2
│   ├───file1
│   ├───file2
│   └───file3
├───dir3
│   └───file4
│   └───file5
├───dir4
    └────file6
```

dir1 has no influence on the progress (which can be considered correct, I like this about the new solution)
files in dir2 will set the percentage to

file1 : 25% / 3 -> ~ 8.4
file2: file1 + 25% / 3 -> ~ 16.8
file3: file2 + 25% / 3 -> ~ 25

however we restart setting them in dir3

file4: 25% / 2 -> 12.5%
file5: file4 + 25% / 2 -> 25%

In fact I have run this to double check and this is what each file calls the invoke with:

```
file1.dll: 8.333333
file2.dll: 16.66667
file3.dll: 25
file4.dll: 12.5
file5.dll: 25
file6.dll: 25
```

Reply:
The problem with calculating the root folder size (or the total number of files, which is another way of doing it), is that one has to iterate through everything first, and then iterate again when zipping - which will impact performace and is probably not needed.